### PR TITLE
nvidia_x11: #127693

### LINKED ISSUE #127693: NVIDIA 465 beta driver breaks suspend, due to lacking `nvidia-sleep.sh` script
In

https://github.com/NixOS/nixpkgs/pull/73530/files#diff-98b29c6e49617265eccb2562afa3a70ba277e6af9508d908c001a4ed398b5e48R49

`nvidia-sleep.sh` is installed if it exists, and the install siliently continues if it doesn't exist.

```sh
    # Install systemd power management executables
    if [ -e nvidia-sleep.sh ]; then
        sed -E 's#(PATH=).*#\1"$PATH"#' nvidia-sleep.sh > nvidia-sleep.sh.fixed
        install -Dm755 nvidia-sleep.sh.fixed $out/bin/nvidia-sleep.sh
    fi
```

In the service

https://github.com/NixOS/nixpkgs/pull/73530/files#diff-28ea4da3a9a595f8bd1a37a555c229f998f3a11b0715ebc5842ba266384eba73R247

```
          ExecStart = "${nvidia_x11.out}/bin/nvidia-sleep.sh '${state}'";
```

the script is executed unconditionally, even if the script doesn't exist.

That breaks suspend with the 465 beta driver, because there the script does not exist.

I think we should

* [ ] fail loudly if it doesn't exist (so that if you close the laptop and put it in a bag, it doesn't catch fire)
* [ ] figure out where the script went / what the replacement is.

---

CC @lovesegfault @eadwu

### PR DESCRIPTION
**Pull request by eadwu**
###### Motivation for this change
Tested build of `linuxPackages.nvidia_x11_beta`

Fixes: #127693 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by baracoder** [comment]
Unfortunately this breaks `nvidia_x11_legacy390` since they share `builder.sh`
```
bara@hal ~/src/nixpkgs (git)-[pr-127711] % nix-build -A linuxPackages_4_14.nvidia_x11_legacy390 

...

  LD [M]  /build/kernel/nvidia.ko
make[2]: Leaving directory '/nix/store/chq72rfvi2kx5ajv8phpmlg5m7n9gyjp-linux-4.14.237-dev/lib/modules/4.14.237/build'
make[1]: Leaving directory '/nix/store/chq72rfvi2kx5ajv8phpmlg5m7n9gyjp-linux-4.14.237-dev/lib/modules/4.14.237/source'
installing
Failed to find necessary files for power management for systemd (nvidia-sleep.sh, nvidia)
error: builder for '/nix/store/c19d0yav4fy8ql489z6762d4xi3vnima-nvidia-x11-390.143-4.14.237.drv' failed to produce output path for output 'bin' at '/nix/store/c19d0yav4fy8ql489z6762d4xi3vnima-nvidia-x11-390.143-4.14.237.drv.chroot/nix/store/r4fhfnx9snc8xvbpvavcx1p8h39967v7-nvidia-x11-390.143-4.14.237-bin'
```

The differences to legacy 3xx versions accumulate. Maybe it is time to make them two separate scripts?

**Comment by eadwu** [comment]
Oops sorry, I completely forgot about this, since the files are missing from 390.xx then probably the way to go is just add the assertion outlined in #127693 which I've done, though I don't really like the brittle nature of it.

Concerning the splitting of the script, I'm not really sure how much is split and/or shared between versions though I do agree it's getting pretty ugly. For the `installPhase` problem here, maybe it would work to dynamically construct the `builder.sh` using `substituteAll` with some version bounds.

File: pkgs/os-specific/linux/nvidia-x11/builder.sh [hunk 1]
```diff
@@ -56,6 +56,17 @@ installPhase() {
...
+    ## 465.xx+ redundancies
+    if [ -e systemd/nvidia-sleep.sh ]; then
+        sed -E 's#(PATH=).*#\1"$PATH"#' systemd/nvidia-sleep.sh > nvidia-sleep.sh.fixed
+        install -Dm755 nvidia-sleep.sh.fixed $out/bin/nvidia-sleep.sh
+    fi
+
+    if [ -e systemd/system-sleep/nvidia ]; then
+        sed -E "s#/usr(/bin/nvidia-sleep.sh)#$out\\1#" systemd/system-sleep/nvidia > nvidia.fixed
+        install -Dm755 nvidia.fixed $out/lib/systemd/system-sleep/nvidia
+    fi
```
Comment (veprbl) [inline pkgs/os-specific/linux/nvidia-x11/builder.sh:68]: This has landed in 3f4bb8ff5a8da1d5b2407e0d6cfb2acfc5bb7f90

**Comment by vcunat** [comment]
> Thanks for the assertion, that's great!

I suspect it needs build-during-evaluation, which is a problematic thing to use in NixPkgs... though at least this shouldn't be weighting on Hydra due to being unfree anyway.

**Comment by nh2** [comment]
I believe a build-time assertion on the nvidia package would also do it, as long as it doesn't silently fail as before.

**Pull request merged**